Support runtime programmatic set up for basic and form-based authentication using fluent API

michalvavrik opened the pull request:
- this PR covers portion of https://github.com/quarkusio/quarkus/issues/16728
- simplifies the `HttpSecurityConfiguration` getter in the light of the https://github.com/quarkusio/quarkus/pull/48500#issuecomment-2996128324 comment
- any `HttpAuthenticationMechanism` can be registered programmatically with this the API introduced in this PR

sberyozkin commented on docs/src/main/asciidoc/security-authentication-mechanisms.adoc:
```diff
@@ -217,6 +217,32 @@ The following properties can be used to configure form-based authentication:
+[[form-based-auth-programmatic-set-up]]
+==== Set up Form-based authentication programmatically
+
+In addition to the configuration properties listed above, Quarkus supports programmatic set up during the runtime as in the example below:
```
```suggestion
In addition to the configuration properties listed in the <<form-configuration-properties>> section, Quarkus supports programmatic set up during the runtime as in the example below:
```

where the `form-configuration-properties` is an anchor for that section

sberyozkin commented on docs/src/main/asciidoc/security-basic-authentication.adoc:
```diff
@@ -69,6 +69,33 @@ For more information about how you can secure your Quarkus applications by using
+}
+----
+<1> Enable basic authentication.
+Downside is that Quarkus doesn't know that the basic authentication is required during the build-time.
```
This probably should be moved to the `NOTE`, and instead of using a `downside` word,  clarify that using the build time configuration is needed for some cases like generating an OpenAPI schema, etc

sberyozkin commented:
@michalvavrik It looks very good, thanks for keeping adding these enhancements iteratively, makes it much easier to review...

What I've been wondering about is this: so users can set up Basic and Form programmatically, but this setup can not be complete for example without either JPA or JDBC extensions or simple mapping supported by the file properties.
I.e, the idea with the programmatic setup is that users can avoid configuring application properties...

What are your thoughts about it, should we add the identity provider builder capability, where you can transition to `jdbcItentity()` or `jpaIdentity()` to complete ? Or may be better have events for `JpaIdentity`, `JdbcIdentity` ?

I'm fine with having it done in another PR, let's open an issue to discuss options ?

michalvavrik commented:
> @michalvavrik It looks very good, thanks for keeping adding these enhancements iteratively, makes it much easier to review...
> 
> What I've been wondering about is this: so users can set up Basic and Form programmatically, but this setup can not be complete for example without either JPA or JDBC extensions or simple mapping supported by the file properties. I.e, the idea with the programmatic setup is that users can avoid configuring application properties...
> 
> What are your thoughts about it, should we add the identity provider builder capability, where you can transition to `jdbcItentity()` or `jpaIdentity()` to complete ? Or may be better have events for `JpaIdentity`, `JdbcIdentity` ?

Apologies, I just deferred solving this because I don't know **yet** how to. The "issue" (and a good design) is that there is no direct binding between the `HttpAuthenticationMechanism` and `IdentityProvider`. With current structure, we cannot guarantee that some provider will be used for some mechanism, so it needs to be independent, otherwise it will be confusing. We have 2 options:

1. refactor current code to allow to select some identity provider for the mechanism (it can be done, but there are downsides like tight coupling)
2. Provide independent builder (probably the event again).

I think having `form.jpaProvider()` or `form.jpaIdentity()` is very nice and would probably cover most users scenarios as many users don't care about generic case where they can just switch extensions and runtime properties and it will still work. But it also requires the mechanisms to be aware of builtin identity providers like JPA or JDBC while they can't really bring these extensions in, nor to configure them.

What we could do would be `form.identityProvider(jpaProvider());` while also having `basic.identityProvider(elytronProperties())` and provide some util methods that allows to build the identity providers inside these individual extensions (JPA, elytron, ...).

I think it only has value for a situation when users have multiple identity providers, otherwise they don't need to specify concrete identity provider on the form/basic auth. And how many users will have multiple identity providers for same authentication request type (like username password)?

So personally I'd have separate event to keep it simple, but I don't have strong opinion, I don't think it is difficult to implement `add the identity provider builder capability, where you can transition to `jdbcItentity()` or `jpaIdentity()` to complete`.

> I'm fine with having it done in another PR, let's open an issue to discuss options ?

I am not sure what is scope for the https://github.com/quarkusio/quarkus/issues/16728, I was thinking the WebAuthn is the only left, but given how little I am familiar with current implementation, we could just close the https://github.com/quarkusio/quarkus/issues/16728 and let current fluent API settle. Maybe we will get feedback and users will ask for it later. WDYT? I would be fine to have a new issue for a fluent API for identity providers.

michalvavrik commented:
In regards of the https://github.com/quarkusio/quarkus/issues/16728, sorry, I just remembered I also meant to add something for mTLS, I already checked and it could be possible, but it will require Clement to agree with, so I don't know yet his thoughts. Let's keep the https://github.com/quarkusio/quarkus/issues/16728 open for now.

sberyozkin commented:
Hi Miichal, np at all, yes, we can look at the identity provider a bit later, I'll keep your comments in mind and get back to them.

So, as far as `basic` and `form` are concerned, we can now configure with individual startup events - which works especially great with `@Authenticated`. 

Now, let's say we have path based programmatic setups that you made possible, like `httpSecurity.paths(...).basic()` etc, and to setup the actual basic or form or oidc authentication, users would have to register event handlers.

I wonder, can we also somehow do it all in one place, for example configure form authentication and set which paths it applies on ? I think this may be something you had in mind but I might've proposed to do events first... 

By the way, we also CORS and CSRF that may be worth looking at.

michalvavrik commented:
> Hi Michal, np at all, yes, we can look at the identity provider a bit later, I'll keep your comments in mind and get back to them.
> 
> So, as far as `basic` and `form` are concerned, we can now configure with individual startup events - which works especially great with `@Authenticated`.

+1

> Now, let's say we have path based programmatic setups that you made possible, like `httpSecurity.paths(...).basic()` etc, and to setup the actual basic or form or oidc authentication, users would have to register event handlers.
> I wonder, can we also somehow do it all in one place, for example configure form authentication and set which paths it applies on ? I think this may be something you had in mind but I might've proposed to do events first...

Nothing really stops us from supporting something like:

```
httpSecurity.path("/basic").basic(Basic.realm("TestRealm").identityProvider(jpaProvider()));
httpSecurity.path("/form").form(Form.loginPage("/custom-login").errorPage("/err").identityProvider(jdbcProvider()));
```

Implementation wise it is not an issue, but agreeing on the design is what slows this down :disappointed: , I for once am not sure what really looks good. In my eyes, I'd go with something like:

```
httpSecurity.path("my-form-path").form(config -> errorPage("error-page").identityProvider(jpaIdentity()).roles("admin");
httpSecurity.path("basic").basic(jdbcIdentity());
httpSecurity.path("other").basic(new CustomIdentityProvider<>()).roles("other");
```

I don't believe users normally need it, so what we have is OK, but allowing to specify the provider would be a nice feature considering you can create instance without CDI, some people might like it. Trouble is what to do for generic authentication mechanism? We can support this for basic and form, but doing more would be more difficult due to the context propagation. I'd not do it unless we know there is a user case.

> By the way, we also CORS and CSRF that may be worth looking at.

agreed, alright, adding it on my list

sberyozkin commented on docs/src/main/asciidoc/security-authentication-mechanisms.adoc:
```diff
@@ -194,29 +194,58 @@ code to destroy the cookie.
+
+}
+----
+<1> Observe the `io.quarkus.vertx.http.security.event.Form` CDI event and customize the configuration provided in the `application.properties` file.
```
```suggestion
<1> Observe the `io.quarkus.vertx.http.security.event.Form` CDI event and configure Form authentication mechanism programmatically.
```

I believe the main idea is not to deal with application.properties

michalvavrik replied:
well, I'll apply your suggestion. I thought we should stress what this API can do, which in my eyes unlike the application.properties is:

- use runtime configuration to decide how security setup looks like
- customize some common config in the app properties in individual deployments

So it can do more. But what you suggest doesn't say it can't do it, so I agree.

sberyozkin commented on docs/src/main/asciidoc/security-basic-authentication.adoc:
```diff
@@ -69,6 +69,33 @@ For more information about how you can secure your Quarkus applications by using
+[[basic-auth-programmatic-set-up]]
+== Set up Basic authentication programmatically
+
+The `io.quarkus.vertx.http.security.event.Basic` CDI event allows to customize the configuration provided in the 'application.properties' during the runtime like in the example below:
```
```suggestion
The `io.quarkus.vertx.http.security.event.Basic` CDI event allows to configure the Basic authentication mechanism programmatically like in the example below:
```

Same suggestion as with the Form, even if it also works in the customization case as you suggest

sberyozkin commented on docs/src/main/asciidoc/security-basic-authentication.adoc:
```diff
@@ -69,6 +69,33 @@ For more information about how you can secure your Quarkus applications by using
+<1> Enable the Basic authentication.
+<2> Use the `QuarkusTestRealm` authentication realm.
+
+NOTE: If you generate an OpenAPI schema document with the SmallRye OpenAPI extension, the Basic authentication must be enabled during the build-time with the `quarkus.http.auth.basic` configuration property.
```
It should probably be:
```
[NOTE]
====
...
====
````

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/BasicImpl.java:
```diff
@@ -0,0 +1,55 @@
+
+import io.quarkus.vertx.http.security.event.Basic;
+
+final class BasicImpl implements Basic {
```
Michal, should we make it more consistent with https://quarkus.io/guides/security-openid-connect-multitenancy#programmatic-startup, as far using the builder is concerned ?

`Basic` interface would probably have `enable()` shortcut which would produce a `BasicBuilder`, and `realm` shortcut that would be doc-ed to also enable it, as well as a plain `bulder()`, something like that... But users are calling `.build()` and then the runtime collecting the instance somehow...

That should also align well with configuring `basic/form` without using events where users would call `build()` I guess.

Thoughts ?

michalvavrik replied:
> realm shortcut that would be doc-ed to also enable it

I actually had it in previous versions, but I dropped it because I realized that it could technically be that some code (or even an extension) configures the basic for users, but doesn't want to enable the mechanism. I think it's not important scenario, so let's do it (enable the basic with "realm")

sberyozkin replied:
Hi @michalvavrik Yeah, on one hand having `Basic.create(basic.builder()....build())` looks a bit noisy but then we would be able to do, outside of the event. `httpSecurity.path("basic").basic(Basic.realm("TestRealm").build())` etc (though I'd advocate for the actual config be done before path here, and refer to whatever is configured with `.path(...).basic()` which we can discuss later in any case)

Thanks

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/security/HttpSecurity.java:
```diff
@@ -78,7 +78,31 @@
+     * @param basic {@link Basic} instance that replaces the configuration in the 'application.properties' file
+     * @return HttpSecurity
+     */
+    HttpSecurity mechanism(Basic basic);
```
Hmm, my earlier comment was assuming it was `mechanism(HttpAuthenticationMechanism)`

michalvavrik replied:
That can be done easily, I have following problem with that approach: what if users configure something in the `application.properties` and then do Basic.realm() and it builds `HttpAuthenticationMechanism`? Which configuration are we going to keep? Are we going to detect `BasicAuthMech` instance and don't add the original mechanism? So `Basic` and `Form` are just ways to configure form/basic, they don't built actual mechanism.

michalvavrik replied:
Done, there is validation in place and 2 tests: `FluentApiBasicAuthValidationFailureTest` and `FluentApiFormAuthValidationFailureTest`

michalvavrik commented on docs/src/main/asciidoc/security-authentication-mechanisms.adoc:
```diff
@@ -194,29 +194,59 @@ code to destroy the cookie.
+    void configure(@Observes HttpSecurity httpSecurity) {   <1>
+        var builder = Form.builder().httpOnlyCookie().loginPage("/my-login.html").errorPage("/my-error.html");
+        httpSecurity
+                .mechanism(builder.build())
```
Thanks Sergey.

> I believe we should also have, in this case, another option, httpSecurity.mechanism(builder.build()).

That is completely same as what is in the example, can you elaborate what you mean, please?

> When we have @Authenticated, is httpSecurity.mechanism(builder.build()) enough ?

Yes, it is. I know there are not tests (because 1. all changes are in Vert.X HTTP extensions so I added all the tests there and 2. there is no space for doubt it in my mind). I could add some if you have doubts.

FYI you can also do `httpSecurity.mechanism(new AnyHttpAuthenticationMechanism());`

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/security/Basic.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Represents a Basic authentication configuration.
+ */
+public record Basic(String realm, Optional<Boolean> enabled) {
```
I have to admit, I'm not sure right now. I mean, this looks neat, and well thought out, I'm just not sure it works for a generic HttpSecurity level fluent API, as we can't offer the same for other mechanisms. I'll have to think also try to suggest some alternatives and we should invite more reviewers.

michalvavrik replied:
OK, as long as you have this on mind, take your time. I'd prefer if this got in for 3.25, but it's not a big deal and if so, there is still some time.

michalvavrik commented:
> Hi @michalvavrik, I do think you are on the right path, IMHO we just need a little bit more thinking how to offer a consistent API that works for every mechanism...

:+1: 

> IMHO we should treat `HttpSecurity` API as experimental right now, to give us more flexibility to change things

Good idea.

sberyozkin commented on docs/src/main/asciidoc/security-authentication-mechanisms.adoc:
```diff
@@ -194,29 +194,59 @@ code to destroy the cookie.
+        var builder = Form.builder().httpOnlyCookie().loginPage("/my-login.html").errorPage("/my-error.html");
+        httpSecurity
+                .mechanism(builder.build())
+                .path("/custom-config/*").form(builder.loginPage("/custom-login.html").build());  <2>
```
@michalvavrik Michal, if we retain right now path based creation of form or basic or ant other custom mechanisms, with `authenticatedWith` then the API becomes non-minimalistic, and quite ambiguous - one can do `httpMechanism.mechanism()`, the duplicate it with `.paths()....authenticatedWith()`, and users may have to do it multiple times, `authenticatedWith()`, if different paths have different authorization policies but the same mechanism, and we haven't even had a single user requirement for basic or form to support it. 
IMHO this is a mechanism specific multi-tenancy concern how to handle such cases. 

`httpMechanism.mechanism()` that you propose for basic for now is a much better start IMHO, it is a single anchor, and a given registered mechanism can be referred to from various places where it is needed

michalvavrik commented:
hey @sberyozkin , I have removed `Basic` and `Form` specific methods, now there is only `httpSecurity.mechanism(httpAuthenticationmechanism)` and I think it makes sense. It is extendable as we will be able to support programmatic build for other mechanisms as well. I also added a test with `@Authenticated`. Let's have a look when the time is right for you. I'll keep this as draft until you agree this is ready to go.

michalvavrik commented on docs/src/main/asciidoc/security-authentication-mechanisms.adoc:
```diff
@@ -194,29 +194,59 @@ code to destroy the cookie.
+        var builder = Form.builder().httpOnlyCookie().loginPage("/my-login.html").errorPage("/my-error.html");
+        httpSecurity
+                .mechanism(builder.build())
+                .path("/custom-config/*").authenticatedWith(builder.loginPage("/custom-login.html").build());  <2>
```
But `authenticatedWith` has nothing to do with this PR, it is already implemented and this "flexibility" you are referring to is just analogy to "quarkus.http.auth.permission."permissions".auth-mechanism=custom-mechanism`. personally I think doing `httpSecurity.path("my").authenticatedWith(new CustomAuthMech())` than defining transport credentials and globally regisgtering a mechanism.

Honestly, this is much simpler doing it programmatically and it is not a "new concept", it is just a programmatic way of doing things.

Would you settle on removing documentation examples and keeping it there? I have hard time removing something that already works on main and I was very happy it got in.

sberyozkin replied:
@michalvavrik 

>"quarkus.http.auth.permission."permissions".auth-mechanism=custom-mechanism

It is not the same at all as `httpSecurity.path("my").authenticatedWith(new CustomAuthMech())` unfortunately.
The above refers to the custom mechanism implementation registered once and referred to from various places.

>personally I think doing httpSecurity.path("my").authenticatedWith(new CustomAuthMech()) is much easier to understand than defining transport credentials and globally registering a mechanism.

Hmm... I did not realize that it bypasses the way the mechanisms are linked to... I'm getting confused now.
What exactly is `CustomAuthMech()` that allows it to be used without it being a registered mechanism ?

michalvavrik commented:
I'd be fine dropping all `authenticatedWith` examples, or even add some warning to consider `global` mechanism registration @sberyozkin (although there is no practical reason for it), but dropping it is just making programmatic setup worse, users will have to prefer "string" based configuration if they want to select authentication and I just ned to understand why.

sberyozkin commented on docs/src/main/asciidoc/security-authentication-mechanisms.adoc:
```diff
@@ -194,29 +194,59 @@ code to destroy the cookie.
+}
+----
+<1> Observe the `io.quarkus.vertx.http.security.HttpSecurity` CDI event and configure Form authentication mechanism programmatically.
+<2> Configure the `/custom-login.html` login page only for requests with the `/custom-config` sub-paths.
```
@michalvavrik Michal, it is not about removing the documentation. I don't get what is going on

sberyozkin replied:
I've had a look...

sberyozkin commented:
@michalvavrik 

>noone forces you to create a new instance for each path

API should not allow it, that the point

sberyozkin commented:
@michalvavrik OK, I've refreshed it by going through your previous PR, let me think a bit, like I said, overall, I'm happy how it is evolving, I just very weary of various ambiguities in the API in general... I'll get back soon

michalvavrik commented:
@sberyozkin I wonder if there can be a middle way between this, because I don't want this PR about `authenticatedWith` (which is not modified here at all). What I'd suggest is that we seriously limit `authenticatedWith` like this:

- accept `string` as you can do with `quarkus.http.auth.permission."permissions".auth-mechanism`
- accept `HttpAuthenticationMechanism` instance that has class that is:
  - one of whitelisted mechanisms, like basic authentication mechanism (it can be only one, I am fine with not allowing form mechanism)  , or the whitelist will be empty, also fine with me!
  - instances whose class package does not start with `io.quarkus.`

Wouldn't it address all your worries and still allow users to register their own custom mechanisms? If any other non-whitelisted mechanism is passes, we raise an error.

If that is not acceptable, ok, I am willing to drop `authenticatedWith(HttpAuthMechanism)` entirely, but just for record, I don't think this makes things easier....

michalvavrik commented:
I'll drop `authenticatedWith(HttpAuthMechanism)`.

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/security/Basic.java:
```diff
@@ -0,0 +1,34 @@
+ * This class provides a way to create a basic authentication mechanism. The {@link HttpAuthenticationMechanism} created
+ * with this class can be registered using the {@link HttpSecurity#mechanism(HttpAuthenticationMechanism)} method.
+ */
+@Experimental("Remains to be determined if this is the best possible API for users to configure basic authentication")
```
```suggestion
@Experimental("This API is currently experimental and may get changed")
```

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/security/HttpSecurity.java:
```diff
@@ -75,10 +76,19 @@
+     * @param mechanism {@link HttpAuthenticationMechanism}
+     * @return HttpSecurity
+     */
+    HttpSecurity mechanism(HttpAuthenticationMechanism mechanism);
```
@michalvavrik 
I propose to add a few shortcuts here, similarly to how you do it with `HttpPermission` below.

`HttpSecurity basic()`, or `HttpSecurity basic(String realm)` and may be a typed form option `HttpSecurity form(FormAuthenticationMechanism)` - we can do it later, just thinking that for build in mechanisms, we should users add them with some type specific support.

You can add it now or later

sberyozkin commented:
Thanks for your work on this PR, @michalvavrik, I hope we'll have a chance to analyze your idea with multiple instances and figure out how to apply it in various combinations.

michalvavrik committed:
feat(security): introduce fluent api for basic and form auth

michalvavrik commented:
> Is this ready to be merged?

I think so, I realized I'd like to add shutdown task, but it doesn't need to be in this PR (because it is required for previous fluent api changes as well), it will be tiny change and I don't have time right now. Let's get this in and I'll follow-up, I am preparing fluent API for mTLS so I can add it there as well.

geoand commented:
🙏🏽
